src/components/FilterBar/index.tsx

```diff
@@ -0,0 +1,109 @@
+import React, { useCallback, useEffect } from 'react'
+import Stack from '@mui/material/Stack'
+import MultiSelectCheckmarks from '../FilterControls/MultiSelect'
+
+import { Button } from '@mui/material'
+import {
+  MediaFilterActionTypes,
+  MediaFilterState,
+} from '../../types/interfaces/MediaFiltersReducer'
+import MediaTitleSearchInput from '../FilterControls/MediaTitleSearchInput'
+import MediaTypeRadioGroup from '../FilterControls/MediaTypeRadioGroup'
+
+const genreOptions = ['action', 'adventure', 'comedy']
+const yearsOptions = ['1981', '1987']
+
+interface FilterBarProps {
+  dispatch: ({ type, payload }: { type: any; payload?: any }) => void
+  state: MediaFilterState
+}
+
+const FilterBar: React.FC<FilterBarProps> = ({ dispatch, state }) => {
+  const { genres, years, searchText, type } = state
+
+  useEffect(() => {
+    console.log('FILTER BAR🚀 ~ genres, years, searchText, type :', searchText)
+  }, [state])
+
+  const createFilterHandler = <K extends keyof MediaFilterState, T>(
```

Comment:
This is neat in concept, but I have to be honest, it makes it harder to review for me, and thus harder to maintain if I were to develop on this project later. As it stands, we're adding special logic that only applies to one case, which is handleTypeChange. I think you're better off with a separate useCallback for each case, and there's no shame in the fact that most of those callbacks just call a basic dispatch.

This also would allow you to improve type safety for each callback.

Reply:
Totally agree! Readability > fancy!